packages/graph/hash_graph/lib/graph/src/subgraph.rs

```diff
@@ -0,0 +1,141 @@
+pub enum GraphElementIdentifier {
+    // TODO: can we create a new type just to generate a Utoipa line for VersionedURI and then use
+    //  that inside a `#[schema(value_type =` expression?
+    OntologyElementId(VersionedUri),
+    KnowledgeGraphElementId(EntityId),
+    Temporary(LinkId),
+}
+
+// TODO: We have to do this because utoipa doesn't understand serde untagged
+//  https://github.com/juhaku/utoipa/issues/320
+impl ToSchema for GraphElementIdentifier {
+    fn schema() -> openapi::Schema {
+        openapi::OneOfBuilder::new()
+            .item(openapi::Object::with_type(openapi::SchemaType::String))
+            .example(Some(serde_json::json!(
+                "6013145d-7392-4630-ab16-e99c59134cb6"
+            )))
+            .into()
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
+#[serde(rename_all = "camelCase")]
+#[serde(tag = "kind", content = "inner")]
+pub enum Vertex {
+    DataType(PersistedDataType),
+    PropertyType(PersistedPropertyType),
+    LinkType(PersistedLinkType),
+    EntityType(PersistedEntityType),
+    Entity(PersistedEntity),
+    Link(PersistedLink),
+}
+
+// WARNING: This MUST be kept up to date with the enum names and serde attribute, as utoipa does
+// not currently support adjacently tagged enums so we must roll our own:
+// https://github.com/juhaku/utoipa/issues/219
+impl ToSchema for Vertex {
+    fn schema() -> openapi::Schema {
+        let mut builder =
+            openapi::OneOfBuilder::new().discriminator(Some(openapi::Discriminator::new("kind")));
+
+        for (kind, schema) in [
+            ("DATA_TYPE", PersistedDataType::schema()),
+            ("PROPERTY_TYPE", PersistedPropertyType::schema()),
+            ("LINK_TYPE", PersistedLinkType::schema()),
+            ("ENTITY_TYPE", PersistedEntityType::schema()),
+            ("ENTITY", PersistedEntity::schema()),
+            ("LINK", PersistedLink::schema()),
+        ] {
+            builder = builder.item(
+                openapi::ObjectBuilder::new()
+                    .property(
+                        "kind",
+                        // Apparently OpenAPI doesn't support const values, the best you can do is
+                        // an enum with one option
+                        openapi::Schema::from(
+                            openapi::ObjectBuilder::new().enum_values(Some([kind])),
+                        ),
+                    )
+                    .required("kind")
+                    .property("inner", schema)
+                    .required("inner"),
+            );
+        }
+
+        builder.into()
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, ToSchema)]
+#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
+pub enum EdgeKind {
+    /// An [`Entity`] has a [`Link`]
+    HasLink,
+    /// A [`Link`] has an [`Entity`] as its destination
+    HasDestination,
+    /// A [`Link`] or [`Entity`] has a [`LinkType`] or [`EntityType`] as its type, respectively
+    HasType,
```

Comment:
Nit: Wouldn't `IsType` be more descriptive?

Reply:
I think they're fairly comparable, but I think `Is` _could_ be more confusing here because `Person` _is_ an Entity Type. And `Number` _is_ a `Data Type`. (Which does mean something different but is close enough to perhaps cause confusion)